Language-Reference-VBA/articles/718dbcd4-63bc-3a75-fa55-7d1e8c65e8b9.md

```diff
@@ -1,25 +1,150 @@
+    
+    d1.CompareMode = vbBinaryCompare             'set the CompareMode on an empty dictionary
+    
+    d1.Add "a", "Athens"                         'Add some keys and items
+    d1.Add "b", "Belgrade"
+    d1.Add "c", "Cairo"
+    d1.Add "C", "Carthage"                       'no error on a binary compare mode (e.g. c <> C)
+    d1.Key("C") = "d"                            'change the key 
+    Debug.Print d1.Count                         'returns 4
 
+    'Late Binding
+    Dim d2 As Object                             'Declare a variable
+    Set d2 = CreateObject("Scripting.Dictionary")  'Bind the variable
+    
+    d2.CompareMode = vbTextCompare               'non-case sensitive compare model
+    
+    d2.Add "a", "Athens"                         'Add some keys and items
+    d2.Add "b", "Belgrade"
+    d2.Add "c", "Cairo"
+    Debug.Print d2.Item("c")                     'Cairo
+    
+    On Error Resume Next                         'attempt to duplicate a key
+    d2.Add "C", "Carthage"                       'error 457 thrown in text compare mode (e.g. c = C)
+    '457 - This key is already associated with an element of this collection
+    Debug.Print Err.Number & " - " & Err.Description
+    On Error GoTo 0
+    
+    If d2.Exists("C") Then _
+        d2.Item("C") = "Carthage"                'overwrite the original 'c' key's item
+    Debug.Print d2.Item("c")                     'Carthage
+    Debug.Print d2.Count                         'returns 3
+    
+    d2.Item("d") = "Cairo"                       'shorthand method; adds if new or updates item of existing key
+    Debug.Print d2.Item("d")                     'Cairo
+    
+    Debug.Print d2.Count                         'returns 4
+
+```
+
+Use a Scripting.Dictionary's Keys as criteria for the [Range.AutoFilter method (Excel)](https://msdn.microsoft.com/en-us/library/office/ff193884.aspx)
+
+```
+    Dim dict As Object
+    Set dict = CreateObject("Scripting.Dictionary")
+    
+    dict.CompareMode = vbTextCompare
+    dict.Add Key:="red", Item:="anything"
+    dict.Add Key:="blue", Item:="anything"
+    dict.Add Key:="green", Item:="anything"
+    
+    With Worksheets("Sheet1")
+        If .AutoFilterMode Then .AutoFilterMode = False
+        With .Cells(1, 1).CurrentRegion
+            'use the dictionary's unique keys as the criteria array
+            .AutoFilter Field:=2, Criteria1:=dict.Keys, Operator:=xlFilterValues
+        End With
+        'data originating at A1 is filtered on column B for red, blue or green
+    End With
 ```
 
+Access and enumerate the dictionary's keys and items.
+
+```
+    'enumerate by key
+    Dim k As Variant
+    For Each k In dict.Keys
+        Debug.Print k & " - " & dict.Item(k)
+    Next k
+    
+    'enumerate by index position (zero-based index)
+    Dim i As Long
+    For i = 0 To dict.Count - 1
+        Debug.Print dict.Keys(i) & " - " & dict.Items(i)
+    Next i
+
+    'enumerate as elements of an array
```

Comment:
there should be note stating that indexed enumeration can only be used with early binding